auth/auth_test.go

```diff
@@ -0,0 +1,206 @@
+
+		err = key.Set(jwk.KeyUsageKey, "sig")
+		require.NoError(t, err)
+		err = key.Set(jwk.AlgorithmKey, "RS256")
+		require.NoError(t, err)
+		marshalledKey, err := json.Marshal(key)
+		require.NoError(t, err)
+
+		w.Header().Set("Content-Type", "application/json")
+		err = json.NewEncoder(w).Encode(map[string]any{
+			"keys": []json.RawMessage{marshalledKey},
+		})
+		require.NoError(t, err)
+	})
+
+	s := httptest.NewServer(mux)
+	t.Cleanup(func() { s.Close() })
+
+	openidConfig = map[string]any{
+		"issuer":                                s.URL,
+		"authorization_endpoint":                s.URL + "/oauth/authorize",
+		"token_endpoint":                        s.URL + "/oauth/token",
+		"userinfo_endpoint":                     s.URL + "/oauth/userinfo",
+		"jwks_uri":                              s.URL + "/oauth/discovery/keys",
+		"id_token_signing_alg_values_supported": []string{"RS256"},
+	}
+
+	token := jwt.NewWithClaims(
+		jwt.SigningMethodRS256,
+		jwt.MapClaims{
+			"iss":   s.URL,
+			"sub":   subject,
+			"aud":   clientID,
+			"exp":   time.Now().Add(time.Hour).Unix(),
+			"iat":   time.Now().Unix(),
+			"nonce": nonce,
+		},
+	)
+	rawIDToken, err = token.SignedString(rs256)
+	require.NoError(t, err)
+	return s
+}
+
+func TestHandler(t *testing.T) {
+	const (
+		testClientID    = "test-client-id"
+		testCode        = "test-code"
+		testAccessToken = "test-access-token"
+		testSubject     = "018d21f2-0b8d-756a-84f0-13b942a2bae5"
+		testName        = "John Doe"
+		testEmail       = "john.doe@example.com"
+	)
+	someday, err := time.Parse(time.DateOnly, "2021-01-01")
+	require.NoError(t, err)
+	userinfo, err := json.Marshal(
+		map[string]any{
+			"sub":            testSubject,
+			"name":           testName,
+			"email":          testEmail,
+			"email_verified": true,
+			"picture":        "https://example.com/avatar.jpg",
+			"created_at":     someday,
+		},
+	)
+	require.NoError(t, err)
+
+	// Set up the mock service server.
+	mockState := ""
+	mux := http.NewServeMux()
+	mockServiceServer := httptest.NewServer(mux)
+	t.Cleanup(func() { mockServiceServer.Close() })
+	redirectURI := mockServiceServer.URL + "/callback"
+
+	// Set up the mock SAMS server.
+	mockSAMSServer := newMockServer(t, redirectURI, testClientID, testSubject, testCode, testAccessToken, userinfo)
+
+	// Set up auth handlers for the mock service server.
+	h, err := NewHandler(
+		Config{
+			Issuer:         mockSAMSServer.URL,
+			ClientID:       testClientID,
+			ClientSecret:   "test-client-secret",
+			RequestScopes:  []scopes.Scope{scopes.OpenID, scopes.Profile, scopes.Email},
+			RedirectURI:    redirectURI,
+			FailureHandler: DefaultFailureHandler,
+			StateSetter: func(_ http.ResponseWriter, _ *http.Request, state string) error {
+				mockState = state
+				return nil
+			},
+			StateGetter: func(*http.Request) (string, error) {
+				return mockState, nil
+			},
+			StateDeleter: func(http.ResponseWriter, *http.Request) {
+				mockState = ""
```

Comment:
We should ensure that we don't call StateDeleter multiple times in a row, right? e.g.

```go
assert.True(t, mockState != "", "mockState already deleted?")
```

Reply:
Eh, probably not.

I mean, if we called `StateDeleter` somewhere between 1x and 100x times for a given state value that would be weird. Right? But would it be worth the trouble to enforce that we only ever call it 1-time instead of 1 or possibly in some rare cases 2x? No.